### ISSUE TITLE
false positive re: preg_replace

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#654]
I'm only getting this issue locally (i.e. not on getpsalm.org)

when `return preg_replace('/[^a-z]+/', '_', mb_strtolower($this->type));` is used, I get `The declared return type 'string' for Foo::bar' is incorrect, got 'array<mixed, mixed>|string'`.

`preg_replace()` only returns an array if argument 3 is an array.

Additionally, it looks like preg_replace does some funky typecasting when argument 3 is an array such that the return type ends up as `array<int|string, string>`

https://3v4l.org/8Kr1Z
```php
<?php

var_dump(preg_replace(
    '/[a-z]/',
    '-',
    [
        'foo' => 'bar',
        1 => 2,
        1.2 => 3.4,
        1.9 => 8.7,
        'baz' => false,
    ]
));
```

**Comment by muglug** [comment]
@SignpostMarv is there a reason Psalm may not be able to type the third argument as a `string`?

You may want to turn on `totallyTyped` mode just to see if `mixed` is complained about

**Comment by SignpostMarv** [comment]
'tis the same with or without totallyTyped mode.

see: https://github.com/SignpostMarv/daft-db-object/blob/cd6093e4c549fa0c36403e97d5909b6d45357b2a/easydb-src/TestObjectRepository.php#L80

**Comment by muglug** [comment]
Ah, this is being caused by a polyfill overriding `mb_strtolower` with a `mixed` return type.

The solution is to disallow function stubbing for functions that Psalm handles internally (like `mb_strtolower`).

**Comment by weirdan** [comment]
Seen similar problem with `paragonie/random_compat` polyfill (also quite popular), where fallback implementation docblock was overriding the built-in function type.

**Comment by SignpostMarv** [comment]
@weirdan I've had commits accepted to paragonie/easy-db, so if you can pass me a repro gist/mini-project that triggers the issue I could perhaps tinker with it if necessary ?

**Comment by SignpostMarv** [comment]
@muglug fixed with 3fe9a06 , I've also required ext-mbstring for good measure.

**Comment by muglug** [comment]
@weirdan what issues are you seeing with `sodium/random_compat`? All those return types are checked by Psalm...

**Comment by weirdan** [comment]
@SignpostMarv @muglug The problem I had with `random_compat` was that it was declaring `random_bytes()` with a void return type (in docblock). The problem has since [been fixed](https://github.com/paragonie/random_compat/commit/37039f48e7a3f7ee6caa3936c815768029ad1d6e#diff-3389ff8e306f9430cb1a83d3ef20c131R206) in 2.0.12 (in fact it's the only entry in changelog for that release only says "fixed psalm errors for dependents"), but you can tinker with 2.0.11 if you want to try to address this on the psalm's side.

`composer require paragonie/random_compat:2.0.11`

Then the following code
```php
<?php
        echo "0x" . bin2hex(random_bytes(40));
```
Fails with 
```
ERROR: InvalidArgument - src/Q.php:7:29 - Argument 1 of bin2hex expects string, void provided
        echo "0x" . bin2hex(random_bytes(40));
```

**Comment by muglug** [comment]
Ah ok - to that point, @paragonie-scott, Psalm doesn't require an explicit return if an exception is thrown.

But also, what do you think Psalm should do here? Those files are scanned before analysing other projects that require `paragonie/sodium_compat` because they're included in `autoload_files`.

To prevent that, I could look for `if(!is_callable('some_function'))` in that initial scanning step and stop if `some_function` is already defined and callable (ditto for `function_exists`). I still want to allow stubbing of functions in general because Psalm uses that for improving array function signatures.

**Comment by weirdan** [comment]
I think there's also some other ways psalm can improve based on that case:
1. There could be a way to see where the type definition psalm uses comes from.
2. It could better integrate with composer, check platform requirements defined in `composer.json` and understand that code won't be used at all (because of the php version check).
3. It could probably check that all of the alternative implementations of the function are the same (it's less of a problem with classes, as they tend to use interfaces to ensure that).

Psalm checking for function existence might actually be undesirable, as psalm could be running on a CI server where configuration doesn't match exactly the production setup. Also, open source projects are limited in what they can assume about the environment they would be run in, and realistically can't test with all possible configurations.
